README.md
# UnMQ [![Build Status][ci-img]][ci]

<img align="right" width="135" height="95" src="http://postcss.github.io/postcss/logo-leftp.png" title="Philosopher’s stone, logo of PostCSS">

[UnMQ] removes media queries from CSS while preserving rules that match a hard-coded viewport. This can be useful for outputting desktop CSS for older browsers like Internet Explorer 8.

```css
/* before */

body {
    font-size: 12px;
}

@media screen and (max-width: 767px) {
    body {
        font-size: 16px;
    }
}

@media screen and (min-width: 768px) {
    body {
        color: #444;
    }
}

/* after */

body {
    font-size: 12px;
}

body {
    color: #444;
}
```

## Usage

Add [UnMQ] to your build tool:

```bash
npm install postcss-unmq --save-dev
```

#### Node

```js
require('postcss-unmq')({ /* options */ }).process(YOUR_CSS);
```

#### PostCSS

Add [PostCSS] to your build tool:

```bash
npm install postcss --save-dev
```

Load [UnMQ] as a PostCSS plugin:

```js
postcss([
    require('postcss-unmq')({ /* options */ })
]);
```

#### Gulp

Add [Gulp PostCSS] to your build tool:

```bash
npm install gulp-postcss --save-dev
```

Enable [UnMQ] within your Gulpfile:

```js
var postcss = require('gulp-postcss');

gulp.task('css', function () {
    return gulp.src('./css/src/*.css').pipe(
        postcss([
            require('postcss-unmq')({ /* options */ })
        ])
    ).pipe(
        gulp.dest('./css')
    );
});
```

#### Grunt

Add [Grunt PostCSS] to your build tool:

```bash
npm install grunt-postcss --save-dev
```

Enable [UnMQ] within your Gruntfile:

```js
grunt.loadNpmTasks('grunt-postcss');

grunt.initConfig({
    postcss: {
        options: {
            processors: [
                require('postcss-unmq')({ /* options */ })
            ]
        },
        dist: {
            src: 'css/*.css'
        }
    }
});
```

## Options

You can define your own viewport for media queries to be evaluated against. Think of the options as the current state of a device and browser.

```js
require('postcss-unmq')({
    // these are already the default options
    type: 'screen',
    width: 1024,
    height: 768,
    resolution: '1dppx',
    color: 3
})
```

If it’s not defined, `device-width` will be given the value of `width`, and `device-height` will be given the value of `height`. Similarly, `aspect-ratio` will be given the value of `device-width` divided by `device-height`.

[ci]: https://travis-ci.org/jonathantneal/postcss-unmq
[ci-img]: https://travis-ci.org/jonathantneal/postcss-unmq.svg
[Gulp PostCSS]: https://github.com/postcss/gulp-postcss
[Grunt PostCSS]: https://github.com/nDmitry/grunt-postcss
[PostCSS]: https://github.com/postcss/postcss
[UnMQ]: https://github.com/jonathantneal/postcss-unmq


CHANGELOG.md
## 1.0.2 (2015-10-24)

- Updated: Documentation

## 1.0.1 (2015-10-24)

- Updated: Tests and documentation

## 1.0.0 (2015-08-26)

- Added: Initial release


test/test.js
var postcss = require('postcss');
var expect  = require('chai').expect;

var plugin = require('../');

var test = function (input, output, opts, done) {
	postcss([ plugin(opts) ]).process(input).then(function (result) {
		expect(result.css).to.eql(output);

		expect(result.warnings()).to.be.empty;

		done();
	}).catch(function (error) {
		done(error);
	});
};

describe('postcss-unmq', function () {
	it('screen and (min-width: 30em) ... preserve child rules ... default options', function (done) {
		test(
			'@media screen and (min-width: 30em) { a { color: black } } b { color: blue }',
			'a { color: black } b { color: blue }',
			{},
			done
		);
	});

	it('screen and (min-aspect-ratio: 1 ... preserve child rules ... default options', function (done) {
		test(
			'@media screen and (min-aspect-ratio: 1) { a { color: black } } b { color: blue }',
			'a { color: black } b { color: blue }',
			{},
			done
		);
	});

	it('screen and (max-width: 30em) ... remove child rules ... default options', function (done) {
		test(
			'@media screen and (max-width: 30em) { a { color: black } } b { color: blue }',
			'b { color: blue }',
			{},
			done
		);
	});

	it('screen and (max-aspect-ratio: 1) ... remove child rules ... default options', function (done) {
		test(
			'@media screen and (max-aspect-ratio: 1) { a { color: black } } b { color: blue }',
			'b { color: blue }',
			{},
			done
		);
	});

	it('screen and (max-width: 30em) ... preserve child rules ... { width: 320 }', function (done) {
		test(
			'@media screen and (max-width: 30em) { a { color: black } } b { color: blue }',
			'a { color: black } b { color: blue }',
			{
				width: 320
			},
			done
		);
	});

	it('screen and (min-width: 30em) ... remove child rules ... { width: 320 }', function (done) {
		test(
			'@media screen and (min-width: 30em) { a { color: black } } b { color: blue }',
			'b { color: blue }',
			{
				width: 320
			},
			done
		);
	});

	it('screen and (max-aspect-ratio: 1) ... preserve child rules ... { width: 320 }', function (done) {
		test(
			'@media screen and (max-aspect-ratio: 1) { a { color: black } } b { color: blue }',
			'a { color: black } b { color: blue }',
			{
				width: 320
			},
			done
		);
	});

	it('screen and (min-aspect-ratio: 1) ... remove child rules ... { width: 320 }', function (done) {
		test(
			'@media screen and (min-aspect-ratio: 1) { a { color: black } } b { color: blue }',
			'b { color: blue }',
			{
				width: 320
			},
			done
		);
	});
});


test/index.js
var tests = {
	'postcss-unmq': {
		'basic': {
			message: 'supports basic usage'
		},
		'basic:width': {
			message: 'supports "width: 320" option',
			options: {
				width: 320
			}
		}
	}
};

var debug = true;
var dir   = './test/fixtures/';

var fs      = require('fs');
var path    = require('path');
var plugin  = require('../');
var test    = require('tape');

Object.keys(tests).forEach(function (name) {
	var parts = tests[name];

	test(name, function (t) {
		var fixtures = Object.keys(parts);

		t.plan(fixtures.length * 2);

		fixtures.forEach(function (fixture) {
			var message    = parts[fixture].message;
			var options    = parts[fixture].options;
			var warning    = parts[fixture].warning || 0;
			var warningMsg = message + ' (# of warnings)';

			var baseName   = fixture.split(':')[0];
			var testName   = fixture.split(':').join('.');

			var inputPath  = path.resolve(dir + baseName + '.css');
			var expectPath = path.resolve(dir + testName + '.expect.css');
			var actualPath = path.resolve(dir + testName + '.actual.css');

			var inputCSS = '';
			var expectCSS = '';

			try {
				inputCSS = fs.readFileSync(inputPath,  'utf8');
			} catch (error) {
				fs.writeFileSync(inputPath, inputCSS);
			}

			try {
				expectCSS = fs.readFileSync(expectPath,  'utf8');
			} catch (error) {
				fs.writeFileSync(expectPath, expectCSS);
			}

			plugin.process(inputCSS, options).then(function (result) {
				var actualCSS = result.css;

				if (debug) fs.writeFileSync(actualPath, actualCSS);

				t.equal(actualCSS, expectCSS, message);

				t.equal(result.warnings().length, warning, warningMsg);
			});
		});
	});
});


index.js
var extend = require('util')._extend;
var postcss = require('postcss');
var mediaQuery = require('css-mediaquery');

var defaultOpts = {
	type:       'screen',
	width:      1024,
	height:     768,
	resolution: '1dppx',
	color:      3
};

module.exports = postcss.plugin('postcss-unmq', function (opts) {
	opts = extend(extend({}, defaultOpts), opts);

	if (!('device-width' in opts)) {
		opts['device-width'] = opts.width;
	}

	if (!('device-height' in opts)) {
		opts['device-height'] = opts.height;
	}

	if (!('aspect-ratio' in opts)) {
		opts['aspect-ratio'] = opts['device-width'] / opts['device-height'];
	}

	if (!('orientation' in opts)) {
		opts.orientation = opts['aspect-ratio'] >= 1 ? 'landscape' : 'portrait';
	}

	return function (css) {
		css.walkAtRules('media', function (rule) {
			if (mediaQuery.match(rule.params, opts)) {
				rule.replaceWith(rule.nodes);
			} else {
				rule.remove();
			}
		});
	};
});
